### ISSUE TITLE
Support composition of model sets with single models

### ISSUE THREAD

**Issue by jehturner** [astropy/astropy#8238]
I would like to represent wavelength shifts between image rows (eg. IFU fibres) using a 1D model set and chain that with a 2D lambda(x,y) surface fit. This involves treating the y dimension as a model set axis in one model but a co-ordinate axis in the other that it's composed with. (ie. each image row has its own `Shift` model but they all share a `Polynomial2D` or similar for the distortions).

(I think this could also be done with a `gwcs.RegionsSelector`, but that makes less sense for thousands of fibres, one per image row, than it does for a handful of MOS or IFU slits and was not the approach that @nden and I discussed previously.)

The first barrier to piping a 1D model set straight into a 2D single model is simply that the numbers of models are required to match, eg.:
```
> (Shift([0.,0.,1.,0.,0.], n_models=5) & Shift([[0.,0.,0.,0.,0.]], n_models=5, model_set_axis=1)) | Polynomial2D(1, c0_1=0, c1_0=1)
...
ValueError: Model sets must have the same number of components.
```
Note that the same constraint also prevents `Shift(n_models=5) & Identity()` from working, because Identity has no `n_models` parameter -- but that's easy enough to work around using Shift(0.) for now, as shown above.

Simply disabling the check `left._n_models == right._n_models` in `CompoundModelMeta` seems sufficient to get the above example working for testing purposes, *but* I think it only happens to work because the input grid is square; if I try with a (5,4) grid instead, I get this, from `prepare_inputs`:
```
> y, x = np.mgrid[0:5,0:4] 
> (Shift([0.,0.,0.,0.,0.], n_models=5) & Shift([[0.,0.,0.,0.]], n_models=4, model_set_axis=1))(x,y)
...
ValueError: Model input argument 'x0' of shape (5, 4) cannot be broadcast with parameter 'offset_0' of shape (5,).
```

@nden previously suggested using array-valued parameters instead of model sets, like so:
```
> m = Shift([1,2,3,4]) & Shift([0,0,0,0]) | Polynomial2D(1, c0_0=1, c0_1=1, c1_0=1)
> m(1, 1)
Out[160]: array([ 4.,  5.,  6.,  7.])
```
However, when evaluating over an array of co-ordinates (rather than a single point), I seem to run into the same broadcasting problem as for model sets:
```
> y, x = np.mgrid[0:5,0:4]
> (Shift([0.,0.,1.,0.,0.]) & Shift([[0.,0.,0.,0.]]) | Polynomial2D(1, c0_1=0, c1_0=1))(x,y)
ValueError: Model input argument 'x0' of shape (5, 4) cannot be broadcast with parameter 'offset_0' of shape (5,).
```
I think this is probably a detail of the assumptions used when checking inputs for compound models?

It would be great if this use case can be made to work, perhaps by allowing `n_models=N` to match `n_models=1` in composition (a bit like NumPy broadcasting)? I might dig into the problem a bit more later, but am documenting it here in the meantime. I think it would be ideal if this can be made to work with model sets, since those are fittable and @nden says array-valued parameters are not (I think the syntax also ends up being cleaner etc.), but I'm open to whatever works :slightly_smiling_face:. Thanks.

**Comment by perrygreenfield** [comment]
Let me see if I can recast this a bit differently; first in a more abstract way. You would like to combine a function of y that provides a shift in x, and use that with a 2d distortion model. If the 2d distortion model consists of the form u = f(x,y); v = g(x,y) and the shift function is s(y) (understanding it is a discontinuous function of y), then the result desired is 

u = f(x + s(y), y); v = g(x + s(y), y)

Do I understand correctly?

**Comment by perrygreenfield** [comment]
If so, then if you can provide a model s  as a table look up function (I think we have that but if not it can be added) then this would be handled as a compound model closely following the previous expression.

transform = mapping(0, 0, 1) |  (identity(1) + s ) & identity(1) | mapping(0, 1, 0, 1) | ( f & g)

@nden is this right. I'm a novice at this sort of thing :-)

**Comment by perrygreenfield** [comment]
The tabular model would be appropriate. It uses interpolation, and for your use, it is likely you would want to bracket the y values for the y region where the x shift is constant (is that clear?)

**Comment by jehturner** [comment]
> Do I understand correctly?
Yes, I think so (but for the time being we could probably ignore `g()`, since I'm mainly concerned about wavelength calibration here).

**Comment by jehturner** [comment]
I need to pop out for a bit, but this seems promising, at least in this case; thanks @perrygreenfield. Is `Tabular1D` fittable?

**Comment by pllim** [comment]
AFAIK, `Tabular1D` is a lookup table. You can interpolate it, but not sure about fitting.

**Comment by perrygreenfield** [comment]
No, I'm pretty sure it isn't since array elements aren't individual parameters. This is the sort of problem that usually uses a different technique (like hundreds of Shift models :-), but seriously, it is tricky since all values of y don't necessarily get fit. Perhaps some automatic adapter that makes a selected subset of elements in an array fittable. That would take some thought.

**Comment by jehturner** [comment]
I can probably fit a `Shift` model set and then copy the parameters into a `Tabular1D`? It's not esp. elegant but I think it ought to work.

**Comment by jehturner** [comment]
The awkward bit seems to be adding 2 models that take different inputs (identity(x) and shift(y)), since addition creates a compound model over a single set of inputs. What you have above isn't quite right, @perrygreenfield, because the shift table gets fed the `x` input (but hey, you came up with that really fast!). What occurs to me is to use a linear 2D polynomial to combine the x input with (a shifted copy of) the y input, like so (in the absence of distortions etc.):
```
> y, x = np.mgrid[0:5,0:4]
> fitted_shifts = Shift([0.,0.,1.,0.,0.], n_models=5)
> shift_table = Tabular1D(np.arange(0, 5), fitted_shifts.parameters, bounds_error=False, fill_value=None)
> wave = Mapping((0,1,1)) | Identity(1) & shift_table & Identity(1) | Polynomial2D(1, c0_0=0., c1_0=1., c0_1=1.) & Identity(1) | Polynomial2D(1, c0_0=5000., c1_0=0.5, c0_1=0.)
> print(wave(x, y))
[[5000.  5000.5 5001.  5001.5]
 [5000.  5000.5 5001.  5001.5]
 [5000.5 5001.  5001.5 5002. ]
 [5000.  5000.5 5001.  5001.5]
 [5000.  5000.5 5001.  5001.5]]
```
Any better ideas that don't involve so many identities etc.? Is there some more idiomatic way of combining inputs, like a reverse `Mapping((0,1), 2)`? :scream: (Oops, that doesn't conserve flux.) I'll need to think about any real use cases that this solution doesn't cover (ie. whether my feature request is still justified). Thanks!

**Comment by nden** [comment]
@jehturner Would this work for you (assuming your input in x is 3 and shifts are represented with `Const1D`:

```
m = Identity(1) + Const1D([1, 2, 3]) # this applies the shifts in x
m1 = m & Identity(1)  # Concatenate with y
m2 = m1 | Polynomial2D(1)  
m2([1,2,3], [1,2,3])
```

**Comment by jehturner** [comment]
This also produces the same broadcasting error as in my original examples, @nden -- but I think perhaps that's due to a mistake on my part (I had some trouble remembering all the conventions after several months). If I just add an inner dimension then it seems to work (even if it's a bit uglier):
```
wave = Identity(1) + Const1D([[p] for p in fitted_shifts.parameters]) & Identity(1) | Polynomial2D(1, c0_0=5000., c1_0=0.5, c0_1=0.)
```
Or in your example above, the first line becomes: `m = Identity(1) + Const1D([[1], [2], [3]])`.

Of course that means I could just use this minor adaptation of your original suggestion, instead of a summation:
```
wave = Shift([[0],[0],[1],[0],[0]]) & Identity(1) | Polynomial2D(1, c0_0=5000., c1_0=0.5, c0_1=0.)
```
And in fact, this seems to work for a model set too, as long as `_CompoundModelMeta` is hacked to remove the `left._n_models == right._n_models` check:
```
> wave = Shift([[0],[0],[1],[0],[0]], n_models=5) & Identity(1) | Polynomial2D(1, c0_0=5000., c1_0=0.5, c0_1=0.)
> print(wave(x, y))
[[5000.  5000.5 5001.  5001.5]
 [5000.  5000.5 5001.  5001.5]
 [5000.5 5001.  5001.5 5002. ]
 [5000.  5000.5 5001.  5001.5]
 [5000.  5000.5 5001.  5001.5]]
```
So I think this was almost working all along and the only obstacle to composing a model set with a single model is the `n_models` check?

**Comment by jehturner** [comment]
OK, I think the problem with simply removing that `n_models` constraint when composing with single models is that we then have to assign either `n_models=1` or `n_models=N` to the resulting compound model, but it ends up with a mixture of array and scalar parameters. The first case is just the same as using array-valued parameters, while the second one doesn't behave as expected because of that inconsistency. So doing what you originally proposed (but with the extra inner dimension) may be the best compromise for now; it means copying parameters out of a fitted model set, which is a bit of a hack, but having a model set turn into a single model with array-valued parameters might violate the principle of least surprise?? The alternative might be to broadcast / replicate all the scalar parameters to match the model set dimensionality. Anyway, this is a much smaller problem now.
